### PR TITLE
refactor: use retry.StateChangeConf for encryption-at-rest resource.

### PR DESCRIPTION
**Pull request by marcosuma**
## Description

Use `retry.StateChangeConf` instead of the `for loop` to remove technical debt and improve retry strategy.

Link to any related issue(s): https://jira.mongodb.org/browse/INTMDB-1039

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [X] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [X] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [X] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [X] I have added any necessary documentation (if appropriate)
- [X] I have run make fmt and formatted my code

### CODE REVIEW

**Comment by marcosuma** [comment]
Running `terraform apply` on the example `terraform-provider-mongodbatlas/examples/test-upgrade/encryption-at-rest/aws/v101`:

```
mongodbatlas_project.test: Refreshing state... [id=64fecfd3da30680cf9043172]
mongodbatlas_cloud_provider_access_setup.setup_only: Refreshing state... [id=aWQ=:NjRmZWNmZDU5MWRhNjY2Nzg2OTRkNGI2-cHJvamVjdF9pZA==:NjRmZWNmZDNkYTMwNjgwY2Y5MDQzMTcy-cHJvdmlkZXJfbmFtZQ==:QVdT]
aws_iam_role.test_role: Refreshing state... [id=mongo_setup_test_role]
aws_iam_role_policy.test_policy: Refreshing state... [id=mongo_setup_test_role:mongo_setup_policy]
mongodbatlas_cloud_provider_access_authorization.auth_role: Refreshing state... [id=aWQ=:NjRmZWNmZDU5MWRhNjY2Nzg2OTRkNGI2-cHJvamVjdF9pZA==:NjRmZWNmZDNkYTMwNjgwY2Y5MDQzMTcy]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # mongodbatlas_encryption_at_rest.test will be created
  + resource "mongodbatlas_encryption_at_rest" "test" {
      + id         = (known after apply)
      + project_id = "64fecfd3da30680cf9043172"

      + aws_kms_config {
          + customer_master_key_id = (sensitive value)
          + enabled                = true
          + region                 = "US_EAST_1"
          + role_id                = "64fecfd591da66678694d4b6"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

mongodbatlas_encryption_at_rest.test: Creating...
mongodbatlas_encryption_at_rest.test: Creation complete after 1s [id=64fecfd3da30680cf9043172]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

project_id = "64fecfd3da30680cf9043172"
role_name = "mongo_setup_test_role"
role_policy_name = "mongo_setup_policy"
```

File: mongodbatlas/resource_mongodbatlas_encryption_at_rest.go [hunk 1]
```diff
@@ -239,27 +240,41 @@ func resourceMongoDBAtlasEncryptionAtRestCreate(ctx context.Context, d *schema.R
 		encryptionAtRestReq.GoogleCloudKms = expandGCPKmsConfig(gcpC.([]interface{}))
 	}
 
-	for i := 0; i < 5; i++ {
-		_, _, err := conn.EncryptionsAtRest.Create(ctx, encryptionAtRestReq)
+	timeout := d.Timeout(schema.TimeoutCreate)
+	stateConf := &retry.StateChangeConf{
+		Pending:    []string{"RETRY"},
+		Target:     []string{"COMPLETED", "ERROR"},
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_encryption_at_rest.go:246]: [nit] Should we make RETRY/COMPLETE/ERROR constants and use them in all the resources with a similar strategy?

File: mongodbatlas/resource_mongodbatlas_encryption_at_rest.go [hunk 2]
```diff
@@ -239,27 +240,41 @@ func resourceMongoDBAtlasEncryptionAtRestCreate(ctx context.Context, d *schema.R
...
+		Refresh:    resourceMongoDBAtlasEncryptionAtRestCreateRefreshFunc(ctx, d.Get("project_id").(string), conn, encryptionAtRestReq),
+		Timeout:    timeout,
+		MinTimeout: 30 * time.Second,
+		Delay:      1 * time.Minute,
+	}
+	// Wait, catching any errors
+	_, err := stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf(errorCreateEncryptionAtRest, err))
+	}
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_encryption_at_rest.go:256]:
```suggestion
	if _, err := stateConf.WaitForStateContext(ctx); err != nil {
		return diag.FromErr(fmt.Errorf(errorCreateEncryptionAtRest, err))
	}
```

File: mongodbatlas/resource_mongodbatlas_encryption_at_rest.go [hunk 3]
```diff
@@ -239,27 +240,41 @@ func resourceMongoDBAtlasEncryptionAtRestCreate(ctx context.Context, d *schema.R
...
-	for i := 0; i < 5; i++ {
-		_, _, err := conn.EncryptionsAtRest.Create(ctx, encryptionAtRestReq)
+	timeout := d.Timeout(schema.TimeoutCreate)
+	stateConf := &retry.StateChangeConf{
+		Pending:    []string{"RETRY"},
+		Target:     []string{"COMPLETED", "ERROR"},
+		Refresh:    resourceMongoDBAtlasEncryptionAtRestCreateRefreshFunc(ctx, d.Get("project_id").(string), conn, encryptionAtRestReq),
+		Timeout:    timeout,
+		MinTimeout: 30 * time.Second,
+		Delay:      1 * time.Minute,
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_encryption_at_rest.go:250]: the delay was 10 seconds. Any reason why is now 1 minutes?

File: mongodbatlas/resource_mongodbatlas_encryption_at_rest.go [hunk 4]
```diff
@@ -239,27 +240,41 @@ func resourceMongoDBAtlasEncryptionAtRestCreate(ctx context.Context, d *schema.R
...
+}
+
+func resourceMongoDBAtlasEncryptionAtRestCreateRefreshFunc(ctx context.Context, projectID string, conn *matlas.Client, encryptionAtRestReq *matlas.EncryptionAtRest) retry.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		_, response, err := conn.EncryptionsAtRest.Create(ctx, encryptionAtRestReq)
 		if err != nil {
 			if strings.Contains(err.Error(), "CANNOT_ASSUME_ROLE") || strings.Contains(err.Error(), "INVALID_AWS_CREDENTIALS") ||
 				strings.Contains(err.Error(), "CLOUD_PROVIDER_ACCESS_ROLE_NOT_AUTHORIZED") {
 				log.Printf("warning issue performing authorize EncryptionsAtRest not done try again: %s \n", err.Error())
 				log.Println("retrying ")
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_encryption_at_rest.go:269]: We should use a more golang idiomatic way to handle the error: with error.is() https://gosamples.dev/check-error-type/

**Review by andreaangiolillo — COMMENTED** [review]
Left a few comments. Make sure to run the acceptance tests locally

**Pull request closed without merging**